sonixflasher.c

```diff
@@ -9,382 +9,788 @@
+                break;
+            case SN260:
+                printf("260 Detected!\n");
+                USER_ROM_SIZE  = USER_ROM_SIZE_SN32F260;
+                USER_ROM_PAGES = USER_ROM_PAGES_SN32F260;
+                MAX_FIRMWARE   = USER_ROM_SIZE_KB(USER_ROM_SIZE);
+                CS0            = 0;
+                sn32_variant   = SN260;
+                break;
+            case SN240B:
+                printf("240B Detected!\n");
+                USER_ROM_SIZE  = USER_ROM_SIZE_SN32F240B;
+                USER_ROM_PAGES = USER_ROM_PAGES_SN32F240B;
+                MAX_FIRMWARE   = USER_ROM_SIZE_KB(USER_ROM_SIZE);
+                CS0            = 0;
+                sn32_variant   = SN240B;
+                break;
+            case SN280:
+                printf("280 Detected!\n");
+                USER_ROM_SIZE  = USER_ROM_SIZE_SN32F280;
+                USER_ROM_PAGES = USER_ROM_PAGES_SN32F280;
+                MAX_FIRMWARE   = USER_ROM_SIZE_KB(USER_ROM_SIZE);
+                CS0            = 0xFFFF;
+                sn32_variant   = SN280;
+                break;
+            case SN290:
+                printf("290 Detected!\n");
+                USER_ROM_SIZE  = USER_ROM_SIZE_SN32F290;
+                USER_ROM_PAGES = USER_ROM_PAGES_SN32F290;
+                MAX_FIRMWARE   = USER_ROM_SIZE_KB(USER_ROM_SIZE);
+                CS0            = 0xFFFF;
+                sn32_variant   = SN290;
+                break;
+            case SN240C:
+                printf("240C Detected!\n");
+                USER_ROM_SIZE  = USER_ROM_SIZE_SN32F240C;
+                USER_ROM_PAGES = USER_ROM_PAGES_SN32F240C;
+                MAX_FIRMWARE   = USER_ROM_SIZE_KB(USER_ROM_SIZE);
+                CS0            = 0xFFFF;
+                sn32_variant   = SN240C;
+                break;
+            default:
+                printf("\n");
+                fprintf(stderr, "ERROR: Unsupported bootloader version: %d, we don't support this chip.\n", data[9]);
+                sn32_variant = 0;
+                break;
+        }
+
+        return sn32_variant;
+    } else {
+        fprintf(stderr, "ERROR: Unsupported family version: %d, we don't support this chip.\n", data[8]);
+        return 0;
     }
+}
 
+bool sn32_check_isp_code_option(unsigned char *data) {
+    uint16_t received_code_option = (data[12] << 8) | data[13];
+    printf("Expected Code Option Table: 0x%04X\n", code_option);
+    printf("Received Code Option Table: 0x%04X\n", received_code_option);
+    if (received_code_option != code_option) {
+        printf("Updating Code Option Table from 0x%04X to 0x%04X\n", code_option, received_code_option);
+        code_option = received_code_option;
+        return false;
+    }
     return true;
 }
 
-int hid_get_feature(hid_device *dev, unsigned char *data)
-{
-    return hid_get_feature_report(dev, data, RESPONSE_LEN + 1);
+uint16_t sn32_get_cs_level(unsigned char *data) {
+    cs_level             = 0;
```

Comment:
Docs in the Code Option table refer to this as Code Security bits, so cs value reflects it. Perhaps a naming cleanup is needed. Currently not giving the option to the user to select CS mode, since that can result in ISP mode disable and no access to the bootloader - enforced CS0 avoids bricks